autoload/ale/autocmd.vim

```diff
@@ -1,77 +1,62 @@
+            endif
+
             " Handle everything that needs to happen when buffers are entered.
             autocmd BufEnter * call ale#events#EnterEvent(str2nr(expand('<abuf>')))
-        endif
-        if g:ale_enabled && g:ale_lint_on_enter
-            autocmd BufWinEnter,BufRead * call ale#Queue(0, 'lint_file', str2nr(expand('<abuf>')))
-            " Track when the file is changed outside of Vim.
-            autocmd FileChangedShellPost * call ale#events#FileChangedEvent(str2nr(expand('<abuf>')))
-        endif
-    augroup END
-
-    augroup ALERunOnFiletypeChangeGroup
-        autocmd!
-        if g:ale_enabled && g:ale_lint_on_filetype_changed
-            " Only start linting if the FileType actually changes after
-            " opening a buffer. The FileType will fire when buffers are opened.
-            autocmd FileType * call ale#events#FileTypeEvent(
-            \   str2nr(expand('<abuf>')),
-            \   expand('<amatch>')
-            \)
-        endif
-    augroup END
-
-    augroup ALERunOnSaveGroup
-        autocmd!
-        autocmd BufWritePost * call ale#events#SaveEvent(str2nr(expand('<abuf>')))
-    augroup END
+            if g:ale_lint_on_enter
+                autocmd BufWinEnter,BufRead * call ale#Queue(0, 'lint_file', str2nr(expand('<abuf>')))
+                " Track when the file is changed outside of Vim.
+                autocmd FileChangedShellPost * call ale#events#FileChangedEvent(str2nr(expand('<abuf>')))
+            endif
 
-    augroup ALERunOnInsertLeave
-        autocmd!
-        if g:ale_enabled && g:ale_lint_on_insert_leave
-            autocmd InsertLeave * call ale#Queue(0)
-        endif
-    augroup END
+            if g:ale_lint_on_save
+                autocmd BufWritePost * call ale#events#SaveEvent(str2nr(expand('<abuf>')))
+            endif
 
-    augroup ALECursorGroup
-        autocmd!
-        if g:ale_enabled && g:ale_echo_cursor
-            autocmd CursorMoved,CursorHold * call ale#cursor#EchoCursorWarningWithDelay()
-            " Look for a warning to echo as soon as we leave Insert mode.
-            " The script's position variable used when moving the cursor will
-            " not be changed here.
-            autocmd InsertLeave * call ale#cursor#EchoCursorWarning()
-        endif
-    augroup END
+            if g:ale_lint_on_filetype_changed
+                " Only start linting if the FileType actually changes after
+                " opening a buffer. The FileType will fire when buffers are opened.
+                autocmd FileType * call ale#events#FileTypeEvent(str2nr(expand('<abuf>')), expand('<amatch>'))
+            endif
 
-    if !g:ale_enabled
-        augroup! ALERunOnTextChangedGroup
-        augroup! ALERunOnEnterGroup
-        augroup! ALERunOnInsertLeave
-        augroup! ALECursorGroup
+            " Clean up highlights when buffers are hidden.
+            autocmd BufHidden * call ale#highlight#RemoveHighlights()
```

Comment:
You should probably move these events for the highlights back to the file they are in. It's tricky. They should run if and only if `g:ale_set_highlights` is `1`, but they should also run if the setting is initially off and later turned on while Vim is running.